### PR TITLE
W-10860974: Migrate currently existing class loading creation in Mule Runtime to the new API

### PR DESCRIPTION
**Pull request by rbourbonnavarro**

### CODE REVIEW

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: modules/deployment-model/src/main/java/org/mule/runtime/artifact/activation/internal/TrackingArtifactClassLoaderResolverDecorator.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.mule.runtime.module.artifact.api.descriptor.ArtifactPluginDescriptor;
+import org.mule.runtime.module.artifact.api.descriptor.BundleDescriptor;
+import org.mule.runtime.module.artifact.api.descriptor.DomainDescriptor;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+public class TrackingArtifactClassLoaderResolverDecorator implements ArtifactClassLoaderResolver {
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply modules/deployment-model/src/main/java/org/mule/runtime/artifact/activation/internal/TrackingArtifactClassLoaderResolverDecorator.java:27]: Added

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/builder/DeployableArtifactClassLoaderFactoryProvider.java [hunk 2]
```diff
@@ -46,12 +55,39 @@
...
    */
   public static RegionPluginClassLoadersFactory regionPluginClassLoadersFactory(ModuleRepository moduleRepository) {
-    return new DefaultRegionPluginClassLoadersFactory(new ArtifactPluginClassLoaderFactory(), moduleRepository);
+    return regionPluginClassLoadersFactory(empty(), moduleRepository);
   }
+
+  /**
+   * Creates a new factory for {@link RegionClassLoader}s of Mule Plugins.
+   */
+  public static RegionPluginClassLoadersFactory regionPluginClassLoadersFactory(ArtifactClassLoaderResolver artifactClassLoaderResolver) {
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/builder/DeployableArtifactClassLoaderFactoryProvider.java:89]: Class is new in 4.5

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/builder/DomainClassLoaderBuilder.java [hunk 3]
```diff
@@ -19,12 +19,6 @@
  */
 public interface DomainClassLoaderBuilder {
 
-  /**
-   * @param artifactPluginDescriptors plugin descriptors that will be used by the application.
-   * @return the builder
-   */
-  DomainClassLoaderBuilder addArtifactPluginDescriptors(ArtifactPluginDescriptor... artifactPluginDescriptors);
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/builder/DomainClassLoaderBuilder.java:26]: Class not released yet

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/builder/DomainClassLoaderBuilderFactory.java [hunk 4]
```diff
@@ -21,24 +22,18 @@
  */
 public class DomainClassLoaderBuilderFactory {
 
-  private final DeployableArtifactClassLoaderFactory<DomainDescriptor> domainClassLoaderFactory;
-  private final ArtifactClassLoader parentClassLoader;
-  private final RegionPluginClassLoadersFactory pluginClassLoadersFactory;
+  private final ArtifactClassLoaderResolver artifactClassLoaderResolver;
+  // private final DeployableArtifactClassLoaderFactory<DomainDescriptor> domainClassLoaderFactory;
+  // private final ArtifactClassLoader parentClassLoader;
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/builder/DomainClassLoaderBuilderFactory.java:27]: Done

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/plugin/ArtifactPluginClassLoaderFactory.java [hunk 5]
```diff
@@ -16,6 +16,7 @@
 /**
  * Creates {@link ArtifactClassLoader} for application or domain plugin descriptors.
  */
+// TODO make this a delegate to an ArtifactClassLoaderResolver
```
[the remark this answers is not in the document]
Comment (rbourbonnavarro) [reply modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/api/plugin/ArtifactPluginClassLoaderFactory.java:19]: Added task

**Comment by rbourbonnavarro** [comment]
--verify

**Comment by rbourbonnavarro** [comment]
--validate

**Review by marianogonzalez — APPROVED** [review]

**Comment by rbourbonnavarro** [comment]
--validate

**Pull request merged**